pkg/lib/cockpit/channel.ts

```diff
@@ -0,0 +1,287 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import type { JsonObject } from './_internal/common';
+import { Transport, ensure_transport, transport_globals } from './_internal/transport';
+import { EventEmitter } from './event';
+
+export type ChannelPayload = string | Uint8Array;
+
+export interface BaseChannelOptions extends JsonObject {
+    command?: never;
+    channel?: never;
```

Comment:
Why even making these part of the type then?

Reply:
It's illegal to provide the `command` or `channel` as part of the options array when opening a channel — those come from the channel implementation.  This enforces that the user definitely doesn't pass those things.

Reply:
Sure, but I mean what's the difference of not specifying them at all? If you try to set an unknown field/property of an object, tsc complains just as well?

Reply:
No.  This extends `JsonObject` which has an index type along the lines of `[x: string]?: JsonValue` which means that any property which is not explicitly mentioned is assumed to be possibly present as some JSON value.

If it didn't extend JsonObject then you'd be right (I think)...

Reply:
aaah! Makes sense.